lorder/files/lorder.1
.\"	$NetBSD: lorder.1,v 1.1.1.1 2004/06/11 17:23:55 jeremy-c-reed Exp $
.\"
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)lorder.1	8.2 (Berkeley) 4/28/95
.\"
.Dd April 28, 1995
.Dt LORDER 1
.Os
.Sh NAME
.Nm lorder
.Nd list dependencies for object files
.Sh SYNOPSIS
.Nm
.Ar
.Sh DESCRIPTION
The
.Nm
utility uses
.Xr nm 1
to determine interdependencies in the list of object files
specified on the command line.
.Nm
outputs a list of file names where the first file contains a symbol
which is defined by the second file.
.Pp
The output is normally used with
.Xr tsort  1
when a library is created to determine the optimum ordering of the
object modules so that all references may be resolved in a single
pass of the loader.
.Sh EXAMPLES
.Bd -literal -offset indent
ar cr library.a `lorder ${OBJS} | tsort`
.Ed
.Sh SEE ALSO
.Xr ar 1 ,
.Xr ld 1 ,
.Xr nm 1 ,
.Xr ranlib 1 ,
.Xr tsort 1
.Sh HISTORY
An
.Nm
utility appeared in
.At v7 .


hepmc/ChangeLog
2010-08-18  Lynn Garren

      * src/GenEventStreamIO.cc, StreamHelpers.cc: 
        cleanup unassociated event remnants if corrupt data is found

2010-07-30  Lynn Garren & Andy Buckley

      * src/GenEventStreamIO.cc: create GenCrossSection, PdfInfo, and HeavyIon
        on the stack rather than the heap to fix a memory leak

  --------------------------  HepMC-2.06.01  --------------------------
2010-06-01  Lynn Garren

      * HepMC/GenRanges.h: remove ConstGenVertexParticleRange since 
        there is neither a GenVertex::particle_const_iterator nor a 
        GenVertex::vertex_const_iterator 
      * HepMC/GenVertex.h: remove unimplemented 
        ConstGenVertexParticleRange particles()


  --------------------------  HepMC-2.06.00  --------------------------
  --------------------------  HepMC-2.06.00.b02  --------------------------
2010-05-13  Lynn Garren

      * HepMC/Version.h: version() now takes an optional output stream

      * HepMC/GenEvent.h: add helper method for reading weight names
      * src/GenEventStreamIO.cc: use new helper method to read weight names
      * testHepMCVarious.input: change order of some event info lines
      * test/testIOGenEvent.input: add named weights for testing

2010-05-12  Lynn Garren

      * "make check" now works if momentum units are MEV

2010-05-11  Lynn Garren

      * rename HepMC/GenEventIterators.h to HepMC/GenRanges.h
      
      * src/GenEventStreamIO.cc: fix a bug reading unnamed (old style) weights
      
      * test/testHepMC: add more weight testing

  --------------------------  HepMC-2.06.00.b01  --------------------------

2010-04-05  Lynn Garren

      * HepMC/StreamInfo.h: add reading_event_header flag for use by streaming input
      * src/GenEventStreamIO.cc: Input no longer expects a fixed order of 
        header lines and will ignore lines that are not of a known type.

      * HepMC/HerwigWrapper.h: add hwbeam
      * rename PythiaWrapper6_2.h to PythiaWrapper6_4.h 
        Add structs for more common blocks.

2010-04-02  Lynn Garren

      * HepMC/IteratorRange.h: put the enum in its own header
      * HepMC/GenEvent.h: add vertex_range() and particle_range()
      * HepMC/GenParticle.h: add particles_in(range) and particles_out(range)
      * HepMC/GenVertex.h: add particles(range), 
        particles_in(GenParticle,range), and particles_out(GenParticle,range)
      * test/testHepMCIteration.cc: use the new methods

2010-03-20  Lynn Garren

      * HepMC/GenEventIterators.h: add ConstGenVertexParticleRange,
        ConstGenParticleProductionRange, and ConstGenParticleEndRange
      * HepMC/WeightContainer.h: method to check for weight name is has_key()
      * src/GenEventStreamIO.cc: use quotes to delineate weight names

2010-02-08  Lynn Garren

      * remove unused PythiaWrapper5_720.h, PythiaWrapper6_152.h, and PythiaWrapper6_152_WIN32.h
      * combine HerwigWrapper6_4.h and HerwigWrapper.h in a single file
      * HepMC/PythiaWrapper.h: add getPythiaCrossSection function
      * HepMC/HerwigWrapper.h: add getHerwigCrossSection function
      * fio/PythiaWrapper.cc: implement getPythiaCrossSection function
      * fio/HerwigWrapper.cc: implement getHerwigCrossSection function
      * remove redundant examples/HerwigHelper.h.

2010-02-04  Lynn Garren

      * doc/Makefile.am, examples/Makefile.am: use pkgdatadir to define the instation directory
        The default installation directory is now ${prefix}/share/HepMC/doc
        Override with --datadir=/some/directory/path

      * HepMC/Polarization.h: add is_defined() and set_undefined()
        is_defined returns false if the polarization is undefined
	set_undefined unsets the defined flag and zeroes the Polarization

2010-01-28  Lynn Garren

      * HepMC/GenEventIterators.h: This new header defines a set of classes
        which provide the appropriate iterators ranges.  They can be used to 
	simplify code and also directly with such utilities as Boost foreach.
	These new classes are 
	GenEventVertexRange, ConstGenEventVertexRange, 
	GenEventParticleRange, ConstGenEventParticleRange, 
	GenVertexParticleRange, 
	GenParticleProductionRange, and GenParticleEndRange
      * test/testHepMCIteration.cc: use the new iterator clases and
        also illustrate the use of std::for_each
 
2010-01-26  Lynn Garren

      * test/testHepMCIteration: illustrate use of std::for_each
 
2010-01-25  Lynn Garren

      * HepMC/WeightContainer.h: add map-like functionality, allowing users to
        name weights, while retaining the original vector interface for 
	complete backwards compatibility
	If a name is not specified for a particular weight, the default
	name is the weight's vector index.
	Also add == and != operators.
      * HepMC/GenEvent.h: add a private method used by the stream I/O
      * src/GenEventStreamIO.cc: add a new line for weight names
        The weight name line is optional when reading input and is 
	written only if the event weights object is not empty.
      * test/testWeights.cc: add tests for WeightContainer
      * test/testHepMC: test I/O of named weights
      * examples/testPythiaCopies.cc: set some arbitrary weights
 
2010-01-09  Lynn Garren

      * removed ParticleDataTable.h and all supporting references.

 --------------------------  HepMC-2.05.01  --------------------------

2010-01-04  Lynn Garren

      * HepMC/GenEvent.h: add write_cross_section(std::ostream&)
      * src/GenEvent.cc: output the cross section (if it exists) when printing an event
      * src/GenVertex.cc, GenParticle.cc: save and restore output stream state
        when calling print methods
	
      * examples: Calculate Pythia and Herwig cross sections using
        getPythiaCrossSection from examples/PythiaHelper.h and 
	getHerwigCrossSection from examples/HerwigHelper.h.
        Set the units explicitly.
      * test/testHepMC.cc.in: call write_cross_section()

2009-12-28  Lynn Garren

       * examples/VectorConversion.h: inline the functions


lorder/files/lorder.sh
#!/bin/sh -
#	$NetBSD: lorder.sh,v 1.1.1.1 2004/06/11 17:23:55 jeremy-c-reed Exp $
#
# Copyright (c) 1990, 1993
#	The Regents of the University of California.  All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. All advertising materials mentioning features or use of this software
#    must display the following acknowledgement:
#	This product includes software developed by the University of
#	California, Berkeley and its contributors.
# 4. Neither the name of the University nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
#	@(#)lorder.sh	8.1 (Berkeley) 6/6/93
#

# If the user has set ${NM} then we use it, otherwise we use 'nm'.
# We try to find the compiler in the user's path, and if that fails we
# try to find it in the default path.  If we can't find it, we punt.
# Once we find it, we canonicalize its name and set the path to the
# default path so that other commands we use are picked properly.

if [ "x${NM}" = "x" ]; then
	NM=nm
fi
if ! type "${NM}" > /dev/null 2>&1; then
	PATH=/bin:/usr/bin
	export PATH
	if ! type "${NM}" > /dev/null 2>&1; then
		echo "lorder: ${NM}: not found" >&2
		exit 1
	fi
fi
cmd='set `type "${NM}"` ; eval echo \$$#'
NM=`eval $cmd`

if [ "x${MKTEMP}" = "x" ]; then
	MKTEMP=mktemp
fi
if ! type "${MKTEMP}" > /dev/null 2>&1; then
	PATH=/bin:/usr/bin
	export PATH
	if ! type "${MKTEMP}" > /dev/null 2>&1; then
		echo "lorder: ${MKTEMP}: not found" >&2
		exit 1
	fi
fi
cmd='set `type "${MKTEMP}"` ; eval echo \$$#'
MKTEMP=`eval $cmd`

# only one argument is a special case, just output the name twice
case $# in
	0)
		echo "usage: lorder file ..." >&2;
		exit ;;
	1)
		echo $1 $1;
		exit ;;
esac

# temporary files
N=`${MKTEMP} /tmp/_nm_.XXXXXX` || exit 1
R=`${MKTEMP} /tmp/_reference_.XXXXXX` || exit 1
S=`${MKTEMP} /tmp/_symbol_.XXXXXX` || exit 1

# remove temporary files on exit
trap "rm -f $N $R $S; exit 0" 0
trap "rm -f $N $R $S; exit 1" HUP INT QUIT PIPE TERM 2>/dev/null || \
	trap "rm -f $N $R $S; exit 1" 1 2 3 13 15

# if the line ends in a colon, assume it's the first occurrence of a new
# object file.  Echo it twice, just to make sure it gets into the output.
#
# if the line has " T " or " D " it's a globally defined symbol, put it
# into the symbol file.
#
# if the line has " U " it's a globally undefined symbol, put it into
# the reference file.
(for file in $* ; do echo $file":" ; done ; $NM -go $*) >$N
sed -ne '/:$/{s/://;s/.*/& &/;p;}' <$N
sed -ne 's/:.* [TDGR] / /p' <$N >$S
sed -ne 's/:.* U / /p' <$N >$R

# sort symbols and references on the first field (the symbol)
# join on that field, and print out the file names.
sort +1 $R -o $R
sort +1 $S -o $S
join -j 2 -o 1.1 2.1 $R $S
